packages/internal/src/Polymorphic/Polymorphic.spec.tsx

```diff
@@ -0,0 +1,326 @@
+
+      const TestComponent = () => {
+        const myRef = React.useRef<HTMLAnchorElement>();
+        testRef = myRef;
+        return (
+          <ExampleComponentForwardRef
+            ref={myRef}
+            as={Wrapper}
+            data-testid="hoc"
+          />
+        );
+      };
+
+      const { getByTestId } = render(<TestComponent />);
+      expect(getByTestId('hoc')).toBeInTheDocument();
+      expect(getByTestId('hoc').tagName.toLowerCase()).toBe('span');
+      expect(wrapperDidRender).toHaveBeenCalled();
+      expect(testRef!).toBeDefined();
+      expect(testRef!.current).toBeDefined();
+    });
+  });
+
+  describe('TSDoc output', () => {
+    describe('Polymorphic', () => {
+      const docs = parseTSDoc('internal/src/Polymorphic');
+
+      test('Docs for Polymorphic are generated', () => {
+        const doc = docs?.find(doc => doc.displayName === 'Polymorphic');
+        expect(doc).not.toBeUndefined();
+        expect(doc!.props).toHaveProperty('AsProp');
+        expect(doc!.props).toHaveProperty(`PolymorphicProps`);
+        expect(doc!.props).toHaveProperty('AriaAttributes');
+        expect(doc!.props).toHaveProperty('DOMAttributes');
+      });
+    });
+
+    describe('Higher-Order components', () => {
+      const docs = parseTSDoc('internal/src/Polymorphic/Example');
+
+      test('Docs for test components are generated', () => {
+        const componentNames = docs?.map(doc => doc.displayName);
+        expect(componentNames).toContain('ExampleComponent');
+        expect(componentNames).toContain('ExampleComponentForwardRef');
+      });
+
+      describe.each(['ExampleComponent', 'ExampleComponentForwardRef'])(
+        'Docs for test components contain the expected props',
+        displayName => {
+          test(`${displayName}`, () => {
+            const doc = docs?.find(doc => doc.displayName === displayName);
+            expect(doc).not.toBeUndefined();
+            expect(doc!.props).toHaveProperty('AsProp');
+            expect(doc!.props).toHaveProperty(`${displayName}Props`);
+            expect(doc!.props).toHaveProperty('AriaAttributes');
+            expect(doc!.props).toHaveProperty('DOMAttributes');
+          });
+        },
+      );
+    });
+  });
+});
+
+/** Test utility functions */
+type WrapperProps = React.ComponentPropsWithoutRef<'span'> & {
+  darkMode?: boolean;
+};
+function makeWrapperComponent(): {
+  Wrapper: React.ForwardRefExoticComponent<
+    WrapperProps & React.RefAttributes<HTMLSpanElement>
+  >;
+  wrapperDidRender: jest.Mock;
+} {
+  const wrapperDidRender = jest.fn();
+
+  const Wrapper = React.forwardRef<HTMLSpanElement, WrapperProps>(
+    ({ children, ...rest }: WrapperProps, ref) => {
+      wrapperDidRender();
+      return (
+        <span data-testid="wrapper" {...rest} ref={ref}>
+          {children}
+        </span>
+      );
+    },
+  );
+
+  Wrapper.displayName = 'Wrapper';
+
+  return {
+    Wrapper,
+    wrapperDidRender,
+  };
+}
```

Comment:
Can we add a test that confirms this new implementation plays well with Emotion?

```
// We don't want TS to throw an error here
styled(Button)`
  // styles
`
```